### PR TITLE
Calculate code intel usagestats

### PR DESCRIPTION
**Pull request by efritz**
This adds the `GetCodeIntelUsageStatistics` method in the usagestats package, which will be exposed via

- graphQL (everything available here)
- pings (a **subset** of the things available here - we will add additional data that could change our data collection policies after further discussion).

### CODE REVIEW

File: cmd/frontend/internal/usagestats/codeintel.go [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+			PreciseHoverStatistics:       &eventStatistics{EventLatencies: &eventLatencies{}},
+			FuzzyHoverStatistics:         &eventStatistics{EventLatencies: &eventLatencies{}},
+			PreciseDefinitionsStatistics: &eventStatistics{EventLatencies: &eventLatencies{}},
+			FuzzyDefinitionsStatistics:   &eventStatistics{EventLatencies: &eventLatencies{}},
+			PreciseReferencesStatistics:  &eventStatistics{EventLatencies: &eventLatencies{}},
+			FuzzyReferencesStatistics:    &eventStatistics{EventLatencies: &eventLatencies{}},
+		})
+	}
+
+	eventNames := map[string]func(p *usagePeriod) *eventStatistics{
```
Comment (unknwon) [inline cmd/frontend/internal/usagestats/codeintel.go]: nit: `eventNames` seems a bit confusing here, where I was expecting it returns an event name (as value) by something (as key).

File: cmd/frontend/internal/usagestats/codeintel.go [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+		"codeintel.hover.precise":       func(p *usagePeriod) *eventStatistics { return p.PreciseHoverStatistics },
+		"codeintel.hover.fuzzy":         func(p *usagePeriod) *eventStatistics { return p.FuzzyHoverStatistics },
+		"codeintel.definitions.precise": func(p *usagePeriod) *eventStatistics { return p.PreciseDefinitionsStatistics },
+		"codeintel.definitions.fuzzy":   func(p *usagePeriod) *eventStatistics { return p.FuzzyDefinitionsStatistics },
+		"codeintel.references.precise":  func(p *usagePeriod) *eventStatistics { return p.PreciseReferencesStatistics },
+		"codeintel.references.fuzzy":    func(p *usagePeriod) *eventStatistics { return p.FuzzyReferencesStatistics },
+	}
+
+	for eventName, getEventStatistic := range eventNames {
+		userCounts, err := db.EventLogs.CountUniqueUsersPerPeriod(ctx, periodType, timeNow().UTC(), periods, &db.CountUniqueUsersOptions{
```
Comment (unknwon) [inline cmd/frontend/internal/usagestats/codeintel.go]: The docstring of `CountUniqueUsersPerPeriod` and `calcStartDate` (the former calls) do not have indication that the `now time.Time` should already be UTC before passing to them. I think there are two approaches here:
1. Document both functions that they're expecting time to be UTC.
2. Convert to UTC inside `calcStartDate` so the caller no longer worries about it.

WDYT?

File: cmd/frontend/internal/usagestats/codeintel.go [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+		})
+		if err != nil {
+			return nil, err
+		}
+
+		for i, uc := range eventCounts {
+			getEventStatistic(activityPeriods[i]).EventsCount = int32(uc.Count)
+		}
+
+		percentiles, err := db.EventLogs.PercentilesPerPeriod(ctx, periodType, timeNow().UTC(), periods, DurationField, DurationPercentiles, &db.EventFilterOptions{
```
Comment (unknwon) [inline cmd/frontend/internal/usagestats/codeintel.go]: Save as above

File: cmd/frontend/internal/usagestats/codeintel.go [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+		WeeklyActivities:  weeklyActivities,
+		MonthlyActivities: monthlyActivities,
+	}, nil
+}
+
+func codeIntelActivity(ctx context.Context, periodType db.PeriodType, periods int) ([]*types.CodeIntelUsagePeriod, error) {
+	if periods == 0 {
+		return []*types.CodeIntelUsagePeriod{}, nil
+	}
+	periods = periods - 1
```
[the remark this answers is not in the document]
Comment (efritz) [reply cmd/frontend/internal/usagestats/codeintel.go]: I updated the event logs interface so this is unnecessary. I think the new code is now cleaner and a bit more self-documenting.

File: cmd/frontend/internal/usagestats/codeintel.go (hunk 2 above)
Comment (efritz) [reply cmd/frontend/internal/usagestats/codeintel.go]: I added documentation here, which seems to be more consistent with the current tests.

File: cmd/frontend/internal/usagestats/codeintel.go (hunk 3 above)
Comment (efritz) [reply cmd/frontend/internal/usagestats/codeintel.go]: Also done.

File: cmd/frontend/internal/usagestats/codeintel.go (hunk 1 above)
Comment (efritz) [reply cmd/frontend/internal/usagestats/codeintel.go]: Renamed.

**Pull request merged**